### PR TITLE
Feature - local packages (folder)

### PR DESCRIPTION
**Pull request by jycouet**
Link to: https://github.com/renovateapp/renovate/issues/1215

If we have in a `package.json` links to some local lib `file:../path/to/folder`
Then le local lib package.json will be copied to the tmp folder to be able to generate the right `yarn lock` file.

This is not working with `tgz` files, only folder reference.

### CODE REVIEW

File: lib/workers/branch/lock-files.js [hunk 1]
```diff
@@ -161,6 +189,28 @@ async function writeExistingFiles(config) {
...
 
+function listLocalLibs(dependencies) {
+  logger.trace(`listLocalLibs (${dependencies})`);
+  const toCopy = [];
+  for (const libName in dependencies) {
+    if ({}.hasOwnProperty.call(dependencies, libName)) {
+      const libVersion = dependencies[libName];
+      if (libVersion.startsWith('file:')) {
+        if (libVersion.endsWith('.tgz')) {
+          logger.error(
```
Comment (rarkins) [inline lib/workers/branch/lock-files.js]: logger.error should be reserved only for serious application errors, i.e. something that requires a `fix:`.
logger.warn should be used when there is "maybe" an application error - not as urgent as error.

In this case, it's a user error, so should be only a logger.info. Instead the way to indicate it to users is via the PR itself.

File: lib/workers/branch/lock-files.js [hunk 2]
```diff
@@ -161,6 +189,28 @@ async function writeExistingFiles(config) {
   }
 }
 
+function listLocalLibs(dependencies) {
+  logger.trace(`listLocalLibs (${dependencies})`);
+  const toCopy = [];
+  for (const libName in dependencies) {
+    if ({}.hasOwnProperty.call(dependencies, libName)) {
+      const libVersion = dependencies[libName];
```
Comment (rarkins) [inline lib/workers/branch/lock-files.js]: Would `for (const [libName, libVersion] of Object.entries(dependencies)) {` work here instead?

File: lib/workers/branch/lock-files.js [hunk 3]
```diff
@@ -161,6 +189,28 @@ async function writeExistingFiles(config) {
...
+      if (libVersion.startsWith('file:')) {
+        if (libVersion.endsWith('.tgz')) {
+          logger.error(
+            `Link to local file '${
+              libVersion
+            }' is not supported. Please do it like: 'file:/path/to/folder'`
+          );
+        } else {
+          toCopy.push(libVersion.substring(5));
```
Comment (rarkins) [inline lib/workers/branch/lock-files.js]: I prefer not to use "magic numbers". instead, so something like `.substring('file:'.length);`

File: lib/workers/branch/lock-files.js [hunk 4]
```diff
@@ -123,6 +123,34 @@ async function writeExistingFiles(config) {
...
+      if (toCopy.length !== 0) {
+        logger.debug(`listOfNeededLocalFiles files to copy: ${toCopy}`);
+        await Promise.all(
+          toCopy.map(async localPath => {
+            const resolvedLocalPath = path.resolve(basedir, localPath);
+            // at the root of local Lib we should find a package.json and yarn will use it to update yarn.lock file
+            const resolvedRepoPath = upath.join(
+              resolvedLocalPath.substring(config.tmpDir.path.length + 1),
+              'package.json'
+            );
```
Comment (rarkins) [inline lib/workers/branch/lock-files.js]: Anything local here needs to be reviewed for malicious intent, or simply a bad accident. e.g. a user configuring too deep a relative path such as `../../../../` which might try to overwrite `/`.

File: lib/workers/branch/lock-files.js [hunk 5]
```diff
@@ -123,6 +123,34 @@ async function writeExistingFiles(config) {
...
+            const fileContent = await platform.getFile(resolvedRepoPath);
+            if (fileContent !== null) {
+              await fs.outputFile(
+                upath.join(resolvedLocalPath, 'package.json'),
```
Comment (rarkins) [inline lib/workers/branch/lock-files.js]: Why is 'package.json' joined here?

File: lib/workers/branch/lock-files.js [hunk 6]
```diff
@@ -123,6 +123,34 @@ async function writeExistingFiles(config) {
...
+                fileContent
+              );
+            } else {
+              logger.error(
```
Comment (rarkins) [inline lib/workers/branch/lock-files.js]: Should be user error, not server-side. We can perhaps make it a warn for now to help us check for any bugs but once this functionality seems solid it should be converted to an info message on server side.

**Review by rarkins — CHANGES_REQUESTED** [review]

File: lib/workers/branch/lock-files.js (hunk 5 above)
Comment (jycouet) [reply lib/workers/branch/lock-files.js]: First I wanted to copy the folder, but I found out that to generate the yarn.lock we just need the package.json of the local lib. => So we get only 1 file! 👍

File: lib/workers/branch/lock-files.js (hunk 4 above)
Comment (jycouet) [reply lib/workers/branch/lock-files.js]: Good point.
I handle it now. 👍 
(+ test)

File: lib/workers/branch/lock-files.js (hunk 6 above)
Comment (jycouet) [reply lib/workers/branch/lock-files.js]: got it

File: lib/workers/branch/lock-files.js (hunk 2 above)
Comment (jycouet) [reply lib/workers/branch/lock-files.js]: Yes it's working, and updated 👍

File: lib/workers/branch/lock-files.js (hunk 1 above)
Comment (jycouet) [reply lib/workers/branch/lock-files.js]: Yes, done

File: lib/workers/branch/lock-files.js (hunk 3 above)
Comment (jycouet) [reply lib/workers/branch/lock-files.js]: Yes, good point

File: lib/workers/branch/lock-files.js (hunk 5 above)
Comment (rarkins) [reply lib/workers/branch/lock-files.js]: That's really good. What about the case where you have the binary files? Are those normally accompanied with a `package.json` beside them?

File: lib/workers/branch/lock-files.js [hunk 7]
```diff
@@ -123,6 +123,41 @@ async function writeExistingFiles(config) {
...
+          toCopy.map(async localPath => {
+            const resolvedLocalPath = path.resolve(basedir, localPath);
+            if (!resolvedLocalPath.startsWith(config.tmpDir.path)) {
+              logger.info(
+                `local lib '${
+                  toCopy
+                }' will not be copied because it's out of the repo.`
+              );
+            }
+            // at the root of local Lib we should find a package.json and yarn will use it to update yarn.lock file
```
Comment (rarkins) [inline lib/workers/branch/lock-files.js]: Comment maybe doesn't need to be yarn-specific? I assume same goes for `npm`?

File: lib/workers/branch/lock-files.js (hunk 5 above)
Comment (jycouet) [reply lib/workers/branch/lock-files.js]: Inside the binary file you have the package.json (not beside).
But I was not able to download and write to disk the binary file! (I sent some hours there!)

I would say that this is a frist step to local package, then if someone request tgz we will enhance...

File: lib/workers/branch/lock-files.js (hunk 7 above)
Comment (jycouet) [reply lib/workers/branch/lock-files.js]: Yes, it should work with npm also.
But I didn't test.

I update the comment

**Comment by jycouet** [comment]
Let me know :)

**Comment by jycouet** [comment]
Hi @rarkin, do you know when you can have a look? I would like to enroll my project soon 👍
Thx

**Comment by rarkins** [comment]
I'm away from home right now, which is why I've been slow to merge. Particularly as I'm still concerned about any possible exploits! Thinking whether to put it behind a feature flag so you can start using it while it remains disabled in my GitHub app

**Comment by rarkins** [comment]
I just realised something actually. We should already be copying every package.json from the repo before we run any install! At least, any with dependencies but it could be tweaked.

**Comment by jycouet** [comment]
I understand your concerns.
- About the flag, should I do a flag: `withLocalFolder` and `false` by default? I like the idea!
- On your second point, you are almost right ^^ I thought about this 2 days ago and tested this...

I have:
* client
  * package.json
* api
  * package.json
* localRepo
  * node_modules
    * lib1
      * package.json

=> here renovate doesn't copy: `localRepo/node_modules/lib1/package.json`.

So I tried:
* client
  * package.json
* api
  * package.json
* localRepo
  * lib1
    * package.json

=> here renovate copy: `localRepo/lib1/package.json`.

And renovate renovates also `localRepo/lib1`! and I would like to not have this.

Actually I would like to not have local repos at all!!! but today I have 4 packages like this, and it will take months to remove them... :(


---
So I go for a flag and it seems reasonable?

**Comment by rarkins** [comment]
I'm glad you tested this, as it matches with what I'd expect.

Right now (before this PR), Renovate copies any `package.json` files that it's renovating. `node_modules` is an ignored string by default, so it makes sense that it wouldn't be copied. And it also makes sense that localRepo/lib1 would be renovated if it wasn't in any ignore string. 

There's one more case where a `package.json` won't get renovated - if it has no dependencies or devDependencies listed.

Anyway this confirms what I had thought and leaves open one more possibility instead of the approach in this PR. We already have the `fileList` for each repo so we copy *all* `package.json` files locally before generating lock files, instead of only the ones we're renovating.

**Comment by jycouet** [comment]
Humm, yes I see your points.
The thing is that I don't want to renovate `package.json` in my `localRepo` folder because I don't have the control on this libs (And they have some dependencies).

So, what should we do?
- We abandon the PR?
  - And I work without merging PR on local repos
  - And I plan to remove my local libs?! (6 months?)
- I add the flag? and you merge this PR?

Just let me know, no issue :)

**Comment by rarkins** [comment]
Let's add the flag so you can use it. Call it `copyLocalLibs`, default to false. Warn in description that it may be deprecated in future in case we change approach.

**Comment by jycouet** [comment]
Done ;)

**Review by rarkins — APPROVED** [review]

**Comment by jycouet** [comment]
Thanks @rarkins 🎉

**Pull request merged**